### DISCUSSION TITLE
How to define HTML files in web_accessible_resources?

### DISCUSSION THREAD

**Discussion by DaPotatoMan** [crxjs/chrome-extension-tools#378] [Q&A] [answered]
I want to add the index.html file in web_accessible_resources. But the file just gets copied (unprocessed).

**Comment by jacksteamdev** [comment, marked as answer]
I remember people raising this question before, and we need to figure out some best practices.

I can think of two use cases from the build tool perspective:
- HTML files that should be copied and web-accessible
- HTML files that should be built and web-accessible

@DaPotatoMan I want to be sure I understand:
- What is your use case?
  - Are you accessing `index.html` from a content script?
- What are you doing with the HTML file?
- Have you tried it without `web_accessible_resources`?

**Reply by filipw01** [reply]
@jacksteamdev I think I just noticed the same issue. When using `vite build` HTML files declared in `vite.config.ts` (like standard vite MPA setup) are not included in `web_accessible_resources` which doesn't let me access it from a content script (I'm using it to create a sidebar in an iframe)

**Reply by DaPotatoMan** [reply]
@jacksteamdev I am trying to load the built HTML file on a web app and communicate with the extension.
If I define the HTML file in `web_accessible_resources` and use `vite build` It works perfectly (HTML file is built). But this doesn't work in `dev`

**Reply by jacksteamdev** [reply]
@filipw01 @DaPotatoMan Thanks for the info! Would the following work for you?

### HTML files declared ONLY in `vite.config.ts`

- processed, supports HMR + optimizations
- IS NOT in `web_accessible_resources`
- use case: supports standalone pages like welcome pages (that shouldn't be web accessible)

### HTML files declared ONLY in `web_accessible_resources`

- same as `publicDir`: file is copied, no processing
- IS in `web_accessible_resources`
- ~~causes full page reload on change~~ not possible since file is static
- use case: supports HTML fragments for vanilla JS projects

### HTML file declared in BOTH `vite.config.ts` and `web_accessible_resources`

- processed, supports HMR + optimizations
- IS in `web_accessible_resources`
- use case: supports iframes in content scripts

We probably need to make a decision tree for the docs (eventually).

**Reply by filipw01** [reply]
Yes, that would work perfectly for me

**Reply by DaPotatoMan** [reply]
Same for me!

**Reply by filipw01** [reply]
@jacksteamdev Is there an ETA on this? This is the last thing blocking migration to MV3 PR in my project. I can try to fix it myself and open a PR if it isn't too complex

**Reply by jacksteamdev** [reply]
@filipw01 I'll take a swing at this today and let you know an ETA.

**Reply by jacksteamdev** [reply]
I _think_ it's working now. :sweat_smile: I've added snapshot tests in #386. I need to devise an e2e test, but the snapshots LGTM!

@filipw01 @DaPotatoMan Can you check out this example repo and LMK if it fits your use cases? 

It's a content script that adds an iframe to Wikipedia: https://github.com/jacksteamdev/web-accessible-html

HMR and build work for me :partying_face: 

![image](https://user-images.githubusercontent.com/23390212/170586565-5a8401f5-79a1-4f67-83ae-937e36892757.png)

**Reply by filipw01** [reply]
Works for me too. Also tested it in my repo and it's working great. Thank you for such a quick fix

**Reply by DaPotatoMan** [reply]
@jacksteamdev Working perfectly! However, when 3rd party cookies are disabled the HTML file does not load.

Error:
```console
Failed to load resource: net::ERR_FILE_NOT_FOUND
```

**Reply by jacksteamdev** [reply]
@DaPotatoMan Interesting; thanks for letting me know! This is an edge case we'll need to note in the docs. 

Disabling 3rd party cookies appears to disable 3rd party service workers in iframes, which is what makes extension page HMR work. This issue seems to be related: [Service worker & Blocked cookies · Issue #1551 · w3c/ServiceWorker](https://github.com/w3c/ServiceWorker/issues/1551)

At least this is limited to extension page HMR in content script iframes. 😅
